### PR TITLE
Add reference cell topology class

### PR DESCRIPTION
**Pull request by garth-wells**
This adds reference cell topology information. It is likely to change, but for now it allows the 'closure dof' functions in the FFC generated code to be removed.

The cell topology class will soon get entity ordering functions for constructing dof maps for different cell types.

### CODE REVIEW

File: cpp/dolfin/fem/ReferenceCellTopology.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+namespace fem
+{
+
+class ReferenceCellTopology
+{
+public:
+  static int dim(CellType cell_type);
+  static const int* num_entities(CellType cell_type);
+  static int num_vertices(CellType cell_type);
```
Comment (chrisrichardson) [inline cpp/dolfin/fem/ReferenceCellTopology.h]: We don't need `num_vertices`, `num_edges` and `num_faces`, can just use `num_entities()[0]` etc.

File: cpp/dolfin/fem/ElementDofLayout.cpp [hunk 2]
```diff
@@ -5,46 +5,82 @@
...
+    _cell = dolfin::CellType::hexahedron;
+  else
+    throw std::runtime_error("Ooops");
+
+  const int* num_entities = ReferenceCellTopology::num_entities(_cell);
+  assert(num_entities);
+
+  // Compute closure entities
+  // [dim, entity] -> closure{sub_dim, (sub_entities)}
+  std::map<std::array<int, 2>, std::map<int, std::set<int>>> entity_closure
```
Comment (chrisrichardson) [inline cpp/dolfin/fem/ElementDofLayout.cpp]: Not sure the second `map` is needed, as sub_dim will be 0, 1, 2 etc. i.e. could just be a vector?

Comment (garth-wells) [reply cpp/dolfin/fem/ElementDofLayout.cpp]: Sure - feel free to push a change.

File: cpp/dolfin/fem/GenericDofMap.cpp [hunk 3]
```diff
@@ -217,13 +216,14 @@ void GenericDofMap::ufc_tabulate_dofs(
       const int num_entity_dofs = entity_dofs[d][i].size();
```
Comment (chrisrichardson) [inline cpp/dolfin/fem/GenericDofMap.cpp]: Maybe use a const reference for `entity_dofs[d][i]` here, as it comes up quite a few times below.

File: cpp/dolfin/fem/ReferenceCellTopology.cpp [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+  case CellType::hexahedron:
+    return hexahedron;
+  default:
+    throw std::runtime_error("Unknown cell type.");
+  }
+
+  return nullptr;
+}
+//-----------------------------------------------------------------------------
+// const int* ReferenceCellTopology::get_entities(CellType cell_type, int d0,
```
Comment (chrisrichardson) [inline cpp/dolfin/fem/ReferenceCellTopology.cpp]: delete commented out code

**Comment by michalhabera** [comment]
This introduces lot of code and logic duplication with `CellType.h/cpp, IntervalCell.h/cpp, ...`
I think old classes should be either removed - or this logic be moved there.

**Pull request merged**